File: 628/628.py
# Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self): 
        self.levels = []
        
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        self.traverse(root, 0)
        return self.levels
      
    # preorder traversal of the tree
    # append each item to it's matching level, creating a new list if necessary
    def traverse(self, node: TreeNode, level) -> None:
        if node is not None:
            if len(self.levels) <= level:
                self.levels.append([])
            self.levels[level].append(node.val)
            self.traverse(node.left, level + 1)
            self.traverse(node.right, level + 1)

File: leetcode/628/628.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[][]}
def level_order(root)
    node_list = []
    traverse_and_fill!(root, 0, node_list)
    node_list
end

# depth-first preorder traversal of binary tree
# updates the node_list values in place
def traverse_and_fill!(node, level, node_list)
    return if node.nil?
    
    # add the value of the current level to the right array
    # creating it if necessary
    node_list << [] if node_list[level].nil?
    node_list[level] << node.val
    
    # recursion goes here
    traverse_and_fill!(node.left, level + 1, node_list)
    traverse_and_fill!(node.right, level + 1, node_list)
end
